Output resource key(s)

shemau opened the pull request:
### Description

Output resource key(s) per #486 
Allow multiple resource key(s) per https://github.com/terraform-ibm-modules/terraform-ibm-cos/issues/527

### Release required?
<!--- Identify the type of release. For information about the changes in a semantic versioning release, see [Release versioning](https://terraform-ibm-modules.github.io/documentation/#/versioning). --->

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [x] Major release (`X.x.x`)

##### Release notes content

Inputs are removed:
 - "create_resource_key" 
 - "generate_hmac_credentials" 
 - "resource_key_existing_serviceid_crn" 
 - "resource_key_name" 
 - "resource_key_role"
 
Input added
 
 ```hcl
 variable "resource_keys" {
  description = "The definition of any resource keys to be generated"
  type = list(object({
    name                      = string
    generate_hmac_credentials = optional(bool, false)
    role                      = optional(string, "Reader")
    service_id_crn            = optional(string, null)
  }))
  default = []
}
 ```
 
 Output added
 
 - "resource_keys" (sensitive)
 
 When upgrading some resource `terraform_data` values are destroyed. These were used to track changes to the inputs that are no longer required and are safe to delete (no data is being deleted).

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

shemau commented:
/run pipeline

shemau commented:
/run pipeline

shemau commented:
The run failed on a know provider issue (5002) creating CBRs, which cascaded into all sorts of CBR errors and then delete issues. A re-run should resolve.

shemau commented:
/run pipeline

shemau commented:
/run pipeline

shemau commented:
/run pipeline

daniel-butler-irl commented on modules/buckets/outputs.tf:
```diff
@@ -1,9 +1,37 @@
 output "buckets" {
   description = "Map of buckets created in the Cloud Object Storage Instance"
-  value       = module.buckets
+  value = merge([for name, bucket in module.buckets : { (name) = {
```
Presumably this is a different format output. I think this makes it a breaking change as any consumers of this output will need code change to adopt the updated bucket module... @ocofaigh

ocofaigh replied:
A breaking change is only when infrastructure gets destroyed. Changed in outputs in not a breaking change. See https://terraform-ibm-modules.github.io/documentation/#/versioning

shemau replied:
It is not a change.

The bucket_resource_key is a new output of the root module.
This code de-constructs the output of the root module and re-constructs it without the new key.
So the output is the same. I applied the `main` branch and changed to `resource-key-output` branch and terraform plan/apply showed the output `buckets` was not changed.

ocofaigh commented on outputs.tf:
```diff
@@ -42,6 +42,12 @@ output "bucket_storage_class" {
   value       = local.bucket_storage_class
 }
 
+output "bucket_resource_key" {
```
resource key is bound to an instance, not a bucket right?

shemau replied:
Yes. The module generates one resource key per bucket. Each bucket defines the COS instance it belongs to, so there may be more than one instance and more than one key.
Resource keys have never been exposed, so if we are going to change this, then now would be the time.
We could only output a resource key when an instance is created. I would have to explore how this impacts the buckets module. Dropping this output.

ocofaigh commented on outputs.tf:
```diff
@@ -42,6 +42,12 @@ output "bucket_storage_class" {
 
+output "bucket_resource_key" {
+  description = "Bucket resource key"
+  value       = ibm_resource_key.resource_key
```
Does this work? Isnt `ibm_resource_key` using a count?

shemau replied:
This does work, if create_resource_key is set to false

```hcl
    "bucket_resource_key": {
      "value": [],
      "type": [
        "tuple",
        []
      ],
      "sensitive": true
```

However a deeper look at the count shows that a resource key is only created when a COS instance is created, so it only contains a key when using the root module. When using the buckets sub module it is not created and thus this PR needs a rework.

It seems reasonable to generate a resource key for accessing the instance/bucket with the bucket information.
- It already includes the instance id and guid
- It already includes s3 endpoints (based on region/location)
The options are to exclude the resource key from the buckets sub module, or create them with the bucket information so that they can be found easily with just the bucket name.

ocofaigh reviewed (changes requested):
We need the output in fscloud submodule too

shemau commented:
The current commit creates on resource key per bucket. The resource key is at the COS instance level.

It is possible to use [COS IAM permissions](https://cloud.ibm.com/docs/cloud-object-storage/iam?topic=cloud-object-storage-iam-bucket-permissions&interface=ui) to restrict users access to individual buckets. 

The documentation suggests there could be (for security first/least access principles, there should be) one (or more) instance service IDs/resource keys existing alongside individual access controls for individual buckets. 

More generally:
[adding/viewing service credentials](https://cloud.ibm.com/docs/account?topic=account-service_credentials&interface=terraform) for any IAM enabled service instance.
[how IAM works](https://cloud.ibm.com/docs/account?topic=account-iamoverview#iamoverview)

shemau commented:
Summary of changes:

- root module: 
  - create resource key(s) based on new map for either existing COS instance (if specified) or for new COS instance
  - output new resources keys
  - no changes to existing properties, still create resource keys, still not output
- buckets module:
  - keep old output structure by removing the new output from the root module
  - no mechanism to create keys based on bucket
- fscloud module
  - support resource key(s), passing to route module
  - output resource key(s) generated by root module
  - no changes to existing properties
- advanced example
  - demonstrate how to create different resource keys associated with unique service ids
  
This is ready for another review, at least to make sure it is heading in the agreed direction @ocofaigh @daniel-butler-irl

shemau commented:
/run pipeline

shemau commented:
Is it the right approach to make the serviceid crn a required parameter?
Currently the fscloud implementation does not even pass the serviceid crn (in the legacy options). This guarantees a new service id is created.

Would be better to remove the legacy options from FSCloud at the risk of deleting a serviceID and resource key? Secure by design would seem to suggest the consumer should link the ID and the key; to be clear what the ID is used for. (?)

shemau commented:
/run pipeline

shemau commented:
/run pipeline

shemau commented:
This works other than the upgrade test at https://github.com/terraform-ibm-modules/terraform-ibm-cos/actions/runs/8437308834/job/23107107412#step:7:4458

Deleted:
TestRunUpgradeExample 2024-03-26T14:15:21Z command.go:185:   - resource "ibm_iam_service_id" "resource_key_existing_serviceid" {

Replaced with:
TestRunUpgradeExample 2024-03-26T14:15:21Z command.go:185:   # ibm_iam_service_id.resource_keys_existing_serviceids[0] will be created

This is `not` an upgrade issue, the example changed.

Deleted:
TestRunUpgradeExample 2024-03-26T14:15:21Z command.go:185:   # module.cos_bucket1.ibm_resource_key.resource_key[0] will be destroyed

Replaced with:
TestRunUpgradeExample 2024-03-26T14:15:21Z command.go:185:   # module.cos_bucket1.ibm_resource_key.resource_keys["cos-upgrade-osgnaf-content-reader-key"] will be created

This is an upgrade issue.

This is not a default case, in that we changed the service id (to an list) and a single key to a list in the example.
In the general case four properties were set (these will need to be removed)

```hcl
variable "create_resource_key" {
  description = "Set as true to create a new resource key for the Cloud Object Storage instance."
  type        = bool
  default     = true
}

variable "generate_hmac_credentials" {
  description = "Set as true to generate an HMAC key in the resource key. Only used when create_resource_key is `true`."
  type        = bool
  default     = false
}

variable "resource_key_existing_serviceid_crn" {
  description = "CRN of existing serviceID to bind with resource key to be created. If null a new ServiceID is created for the resource key."
  type        = string
  default     = null
}

variable "resource_key_name" {
  description = "The name of the resource key to be created."
  type        = string
  default     = "cos-resource-key"
}

variable "resource_key_role" {
  description = "The role you want to be associated with your new resource key. Valid roles are 'Writer', 'Reader', 'Manager', 'Content Reader', 'Object Reader', 'Object Writer'."
  type        = string
  default     = "Manager"
}
```


```hcl
  type = list(object({
    name                      = string
    generate_hmac_credentials = optional(bool, false)
    role                      = optional(string, "Reader")
    service_id_crn            = optional(string, null)
  }))

  resource_keys = [
    {
      name                      = var.resource_key_name || "cos-resource-key"
      role                      = var.resource_key_role || "Manager"
      generate_hmac_credentials = var.generate_hmac_credentials ( || defaults to false as before )
      service_id_crn            = var.resource_key_existing_serviceid_crn ( || defaults to null as before )
    }
  ]
  
  moved {
    from: module.<consumer module name>.ibm_resource_key.resource_key[0]
    to: module.<consumer module name>.ibm_resource_key.resource_keys[`var.resource_key_name` || `"cos-resource-key"`]
  }
```

There are number of slightly different migration scenarios. If you were generating a key and looking it up and using outside of the module you might:
- Just destroy and recreate and use the new one.
- Create a resource_keys block, as above based on the old values when using the root module
- If there are multiple root module calls each one may be different
- If using the buckets module, each bucket would have created a service ID and resource key. This no longer happens.
- If using the fscloud module, each instance and bucket would have created service ID and resource key. This no longer happens.

ocofaigh commented on examples/advanced/main.tf:
```diff
@@ -13,12 +13,12 @@ module "resource_group" {
-resource "ibm_iam_service_id" "resource_key_existing_serviceid" {
-  name        = "${var.prefix}-reskey-serviceid"
+resource "ibm_iam_service_id" "resource_keys_existing_serviceids" {
+  count       = 5
```
to reduce the number of service IDs created in our account, can we just create 1 here re-use it?

shemau replied:
[No](https://github.com/terraform-ibm-modules/terraform-ibm-cos/actions/runs/8451636271/job/23150580839#step:7:3692)

```
Error: AdvancedExample 2024-03-27T12:30:38Z logger.go:66: │ Error: [ERROR] Error creating resource key: The service-id 4b6ed744-a2bd-4e94-a8c6-a929f6a56924 provided is in use by another application. For security reasons, each application must have it's own service-id.  Please submit the request with no service-id specified so a new service-id can be generated. with resp code: {
```

Each resource key requires a service id. They will be created as needed automatically if they are not specified.

ocofaigh reviewed (changes requested):
@shemau just 1 comment - please skip upgrade test on your next commit. I'll make this a major version change with supporting release notes

shemau commented:
/run pipeline

ocofaigh commented:
Oh @shemau seems you can re-ruse the service ID: `Error: [ERROR] Error creating resource key: The service-id 4b6ed744-a2bd-4e94-a8c6-a929f6a56924 provided is in use by another application. For security reasons, each application must have it's own service-id.  Please submit the request with no service-id specified so a new service-id can be generated.`

Can you just change the code to use service ID for one of the resource keys then? Also don't forget you need to skip the upgrade test

shemau commented:
/run pipeline

shemau commented:
/run pipeline

terraform-ibm-modules-ops commented:
:tada: This PR is included in version 8.0.0 :tada:

The release is available on:
- [GitHub release](https://github.com/terraform-ibm-modules/terraform-ibm-cos/releases/tag/v8.0.0)
- `v8.0.0`

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
